### DISCUSSION TITLE
Void Linux version is becoming very old

### DISCUSSION THREAD

**Discussion by xoich** [sile-typesetter/sile#1606] [General]
Hello, the manual says sile is available in Void, but the last available version there is 0.9.5. Maybe the manual needs to be updated?

I think the issue with Void not updating is that since that version many new lua libraries have been added as dependencies, and they all need to be packaged for Void.

**Comment by alerque** [comment]
I made some attempts to bring the Void packaging up to speed a while back, you can see the history in [this PR](https://github.com/void-linux/void-packages/pull/18306). The comments in there about packaging location changes are still relevant, as well as your observation about needing a few more Lua modules to be packaged up.

I would be glad to facilitate a refresh from this side but I do not actually use Void myself and am not familiar enough with its packaging routines to make the necessary changes. If somebody that cat take this on would be willing to jump in and start adding Lua modules and eventually bump the package that would be awesome. If there is anything further I can do whether answering questions about SILE or whatever I'd be glad to help.

Note since the PR linked about the Lua dependencies have evolved a bit more. Off the top of my head the `stdlib` package will *not* be needed but a `cldr` and a `fluent` need to be added to the list.

**Comment by xoich** [comment]
thank you. I might consider packaging for void in the future. Unrelated question, the readme says:

> By default all the required Lua libraries will be downloaded and bundled alongside the SILE the instalation. **If you downloaded a source tarball these dependencies are included**

the seems not to be the case? At least no longer

**Reply by alerque** [reply]
Good question. I had to go digging because I thought they were, but clearly it isn't happening. Looking at history the change seems to be on purpose and happened between v0.10.8 and v0.10.9 in cb4d04b. Clearly I started expecting installs to use `luarocks` to fetch them themselves, but I don't remember what the impetus was for that decision. I suspect it was just too error prone to ship a bunch of sources that may or may not get installed and weren't ours vs. just calling or not calling a command to install them. I'm pretty tempted just to update the docs to reflect the current behavior, but if there is some use case for offline installs that can't fetch Lua dependencies on systems that don't also have them (but do have Lua) and dependencies like Harfbuz, ICU, expat, SSL, etc., then please open an issue and I'll look into why I did that and whether it can be reverted.

**Reply by xoich** [reply]
I'm not sure. Ideally you'd want standalone redistributable binaries à la golang, but if a complex installation can't be avoided, personally I really like the way archlinux is installed: Crucial steps are manual and documentation driven. When I tried to compile sile the first time with make I had many luarocks related errors (cannot find lua include files, missing this and that system library). I'm really glad the manual tells me there is an option to skip the automatic luarocks and gives you a way do it yourself. I would probably make it the default considering how many things can go wrong installing those modules with luarocks, and also give some instructions on how to fix common luarocks issues, since few are familiar with lua and luarocks.

Some things that come to mind:

`luarocks --local` if you don't want to use root

Maybe your system has more than one lua installed and you want to choose one in particular:
`luarocks --lua-version 5.4` (there might a better way as this needs to be added always)

You also need to tell this to ./configure:

`export LUA=/usr/bin/lua5.4`

luarocks couldn't find the lua include dir automatically, set it manually:
`luarocks config variables.LUA_INCDIR /usr/include/lua5.4`

You have to update some environmental variables for things to work, you get those with `luarocks path`

I also installed from void packages all the lua modules that were available on it. I had to remove them from the luarocks file otherwise it would keep installing them anyway. Might want to mention that too.

**Reply by alerque** [reply]
A standalone singe binary is not going to happen, at least not until we get a rust based wrapper going with an embeded Lua interpreter and embeded Lua source files for the user modifiable bits (most of the system). Using Lua gives end users flexibility to modify almost any port of the system easily on a per-document basis, but it comes at the cost of a complex install. That's just the way it is.

At the end of the day I don't think we should be focusing on hand-holding end users through source builds and Lua system installations. We can support that for the few cases where it is needed, but the people that need that are going to all have their own system issues. Almost by definition at this point source installs are for when people have weird system situations we havn't otherwise figured out packaging for. The number of possible moving parts prohibits documenting them all, and documenting even some of them is just going to confuse the issue. For example most of the ones you mention (exporting the LUA_INCDIR etc.) should **not** be necessary on most systems and doing so will only get people in more of a mess than the defaults.

I really think the right answer here is to push for packaging fixes, that's really the only sane way to get complex apps out and making sure they continue to function well as other bits of the system get updated. Most Linux systems will update other system libraries like Harfbuzz and ICU and OpenSSL and others that would require SILE be recompiled (not usually for Harfbuzz, but always for ICU, Lua deps need rebuilding for OpenSSL, etc.). With the exception of single-purpose systems like containers using a system package manager is basically the only sane way to fly.

We have quite a few distros covered now, including meta ones like `nix` that are portable across systems. You ran run SILE easily under `nix` on Void Linux now without fussing with source builds.

Lastly for Luarocks dependencies, it is all or nothing: either provide them all yourself on the system or let SILE download and vendor them all locally. Mixing and matching is explicitly not supported because there are too many pitfalls to mixing trees like that. You're of course welcome to hack it the way you did and it can be made to work, but that's on your head as an upstream project we either want to say "use this full set from this path" or "provide a system that has all these things". I know from experience that half-and-half approach makes it really hard to debug when versions of things are incompatible (as, e.g. happened to us with Penlight a while back).

**Reply by xoich** [reply]
That's fine on all points. I just want to remark that until the standalone binaries, or packaging everywhere, there will be people who'll want/need to compile from source. If you don't already have one of the generic package managers the disk-space cost is gigantic. Not sure about nix, but I tried installing with homebrew and I stopped it when disk usage was climbing above 3G, docker somehow ended up requiring an additional 4.7G, that is too much to ask for trying out sile.

**Reply by alerque** [reply]
Homebrew is a beast. Docker is bulky and only really appropriate in an ecosystem that needs it (like mixing with other CI jobs). Nix is a bit lighter. A base Void install that was at 158M before grew to 214M after installing nix and then to 732M after running SILE (so with a full Nix store with all system dependencies). That's a lot considering native packages would probably be about 20M, but still not prohibitive for giving it a try and 500M is a lot less than 4.7G.

```console
$ xbps-install -S nix
$ nix --extra-experimental-features nix-command --extra-experimental-features flakes run nixpkgs/nixpkgs-unstable#sile -- --version
```

**Reply by alerque** [reply]
If you are interested in helping get Void packages up to speed I'd be happy to collaborate. I just [revived a branch to work on a bump](https://github.com/alerque/void-packages/tree/sile-014). I'm a bit confused whether I should be going with lua54 or lua51 though. Probably 5.4 but that means a few more Lua package bumps.

**Reply by xoich** [reply]
ok, I can try to help. I used lua 5.4 because I heard it's faster but otherwise I don't have a strong opinion about it. Let me know if there is anything specific you want help with.

**Reply by alerque** [reply]
If you want *fast*, you should be using LuaJIT. SILE works fine with LuaJIT (and is substantially faster) with one caveat: it doesn't spit out useful trackebacks on errors that show you where in your document or code it ran into trouble.

Lets move toward Lua54.

Off the top of your head do you know what command I should be useing to compile these template files into packages and install them locally for testing? Or know of a good guide for packagers with the necessary ropes for building/testing a series of packages?

**Reply by alerque** [reply]
I was thinking about it on my drive home today and something else doesn't add up. The Docker image for SILE (which is a bare bones Arch Linux system with all the SILE deps) is 167MB over the wire and 529MB in my image cache uncompressed. That's a long way from 4.7G! The Docker software itself on my system takes up 113MB. Still hefty compared to a 20MB system package, but less than a Gig from end to end. Some of my CI servers have dozens of versions of SILE and the layering keeps the total usage pretty sane.

CaSILE is in the multi-gig range. Are you sure you didn't mix up the SILE image with the CaSILE image?

**Reply by xoich** [reply]
About packaging I'm learning the whole thing myself at the moment.

For the docker space size I tried again, it's 4.4G.

```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb        5.9G   72K  5.6G   1% /media/rw
$ podman pull siletypesetter/sile
Trying to pull docker.io/siletypesetter/sile:latest...
Getting image source signatures
Copying blob cb266927e885 done  
Copying blob 946fdc6bb789 done  
Copying blob 3c00ba91c0a2 done  
Copying blob 8bce96fdf943 done  
Copying blob 9310dd2eb8d2 done  
Copying blob 66f1918ed364 done  
Copying blob 61031681f0cf done  
Copying blob 6dc3e33fb076 done  
Copying blob 4f4fb700ef54 done  
Copying blob 5425d8ce00cc done  
Copying config 3a871220cb done  
Writing manifest to image destination
Storing signatures
3a871220cb7cb470514fb435958dfc30b0062c93c49390800ca16dddd20b27b3
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb        5.9G  4.4G  1.2G  80% /media/rw
```

**Reply by xoich** [reply]
I got the gist of it and I think I've managed to package cliargs, I'll send the template after I polish a few things

**Reply by xoich** [reply]
commited the template https://github.com/alerque/void-packages/commit/f766f5af8fc4f2cd4bf8d64b7de43efd92321c26

**Reply by xoich** [reply]
I followed the packaging in alpine of the same module: https://git.alpinelinux.org/aports/tree/community/lua-cliargs/APKBUILD

Not sure why the did it with luarocks, as in this case it does not seem need.

Maybe for future proofing the template, in case the package eventually needs it.

**Reply by alerque** [reply]
That's looking promising, nice work! It helps having a template using the Luarocks install command. Knowing that I think I can make out how to do most of the other deps. That being said I'm hitting the road for 4 days so if you get any more templates spun up before I get to them so much the better. Any work I do will be reflected in that branch without delay.

**Reply by xoich** [reply]
I've added cosmo, I'm yet not sure what to do about the license file since it's not available.

**Reply by xoich** [reply]
hi @alerque, sorry I haven't had much time in the past days. I fixed errors thrown during the packaging of cosmo and cassowary and now those packages should be fine.

**Comment by Omikhleia** [comment]
Took me some time to find what to look at:
https://voidlinux.org/packages/?arch=x86_64&q=sile

Visibly still in 0.9.5.1

I'd suggest removing support on the README, unless we have someone in the Void Linux community to help with it.
An then close this (old) discussion (or does it need to be transferred to issues?)

**Reply by alerque** [reply]
Yup action required here either from them or us. Issue now at #2035.
